File: web/src/engine/packing.ts
/**
 * Packed-integer encoding of team permutations and gamestates for the TS
 * engine (issue #17). Mirrors drafter/common/packing.py with one deliberate
 * difference: JS numbers only hold 53 safe integer bits, so the gamestate key
 * is repacked to 40 bits (two 20-bit team codes) instead of Python's two
 * 24-bit codes, keeping keys on V8's plain-number fast paths (no BigInt).
 *
 * A team code packs into 20 bits:
 *
 *   bits  0..7  : remaining-players bitmask (bit i set => player i remains)
 *   bits  8..11 : defender    (ROLE_NONE = 0xF when unset)
 *   bits 12..15 : attacker_A
 *   bits 16..19 : attacker_B
 *
 * Python's fourth role nibble (discarded_attacker) is dropped: the solver
 * never stores discard-stage gamestates (their game is decided at the parent
 * select_attackers state), so keys only ever encode the none /
 * select_defender / select_attackers stages. In-flight discard decisions live
 * in decoded DraftState objects, never in keys.
 *
 * A gamestate key is friendly + enemy * 2^20 — at most 40 bits, exact in a
 * float64. The low 20 bits are safe for 32-bit bitwise ops; the full key is
 * split with division (bitwise ops would truncate it).
 */

export const ROLE_NONE = 0xf;

const TEAM_BITS = 20;
export const TEAM_SPAN = 1 << TEAM_BITS;

/** Team code with no roles set (all-role nibbles = ROLE_NONE). */
const NO_ROLES = (ROLE_NONE << 8) | (ROLE_NONE << 12) | (ROLE_NONE << 16);

export function encodeTeam(
  remainingMask: number, defender: number, attackerA: number, attackerB: number,
): number {
  return remainingMask | (defender << 8) | (attackerA << 12) | (attackerB << 16);
}

export function encodeNoneTeam(remainingMask: number): number {
  return remainingMask | NO_ROLES;
}

export function teamMask(code: number): number { return code & 0xff; }
export function teamDefender(code: number): number { return (code >> 8) & 0xf; }
export function teamAttackerA(code: number): number { return (code >> 12) & 0xf; }
export function teamAttackerB(code: number): number { return (code >> 16) & 0xf; }

export function encodeGamestate(friendlyCode: number, enemyCode: number): number {
  return friendlyCode + enemyCode * TEAM_SPAN;
}

export function friendlyCodeOf(key: number): number {
  return key - Math.floor(key / TEAM_SPAN) * TEAM_SPAN;
}

export function enemyCodeOf(key: number): number {
  return Math.floor(key / TEAM_SPAN);
}

const POPCOUNT8 = new Uint8Array(256);
for (let i = 0; i < 256; i++) POPCOUNT8[i] = (i & 1) + POPCOUNT8[i >> 1];

export function popcount8(mask: number): number {
  return POPCOUNT8[mask];
}

/** Ascending indices of the set bits — reproduces the Python engine's
 * "remaining players sorted" canonicalisation. */
export function maskIndices(mask: number): number[] {
  const indices: number[] = [];
  for (let i = 0; i < 8; i++) if (mask & (1 << i)) indices.push(i);
  return indices;
}

/** Players on this team: popcount(remaining) + roles set (mirrors
 * packing.team_permutation_n). */
export function teamN(code: number): number {
  let n = POPCOUNT8[code & 0xff];
  if (teamDefender(code) !== ROLE_NONE) n += 1;
  if (teamAttackerA(code) !== ROLE_NONE) n += 1;
  if (teamAttackerB(code) !== ROLE_NONE) n += 1;
  return n;
}

export type StageName = 'none' | 'select_defender' | 'select_attackers';

export function teamStage(code: number): StageName {
  if (teamAttackerA(code) !== ROLE_NONE) return 'select_attackers';
  if (teamDefender(code) !== ROLE_NONE) return 'select_defender';
  return 'none';
}

File: drafter/common/packing.py
"""Packed-integer encoding of team permutations and gamestates (GitHub issue
#13, B2). Replaces the verbose human-readable string keys.

A team permutation over <=8 players packs into a 24-bit int:

    bits  0..7  : remaining-players bitmask (bit i set => player i remains)
    bits  8..11 : defender           (4 bits, ROLE_NONE = 0xF when unset)
    bits 12..15 : attacker_A         (4 bits)
    bits 16..19 : attacker_B         (4 bits)
    bits 20..23 : discarded_attacker (4 bits)

Indices are per side (friendly = CSV rows, enemy = CSV columns), 0..n-1. A
gamestate key is a single int: the friendly team code in the low 24 bits and the
enemy team code in the next 24 (so it fits a numpy int64 and packs the whole
state into one hashable/array-able value, B3). The draft stage is derivable from
either code (see draft_stage_of_code). Names live only in the presentation layer
(drafter.solver.context NameIndex).
"""
from drafter.common.draft_stage import DraftStage

ROLE_NONE = 0xF

_DEFENDER_SHIFT = 8
_ATTACKER_A_SHIFT = 12
_ATTACKER_B_SHIFT = 16
_DISCARDED_SHIFT = 20
_ROLE_MASK = 0xF
_REMAINING_MASK = 0xFF

# A team code occupies 24 bits; the gamestate key stacks enemy above friendly.
_TEAM_BITS = 24
_TEAM_MASK = (1 << _TEAM_BITS) - 1


def encode_gamestate(friendly_code, enemy_code):
    return friendly_code | (enemy_code << _TEAM_BITS)


def decode_gamestate(key):
    return (key & _TEAM_MASK, key >> _TEAM_BITS)


def _pack_role(index):
    return ROLE_NONE if index is None else index


def _unpack_role(code, shift):
    value = (code >> shift) & _ROLE_MASK
    return None if value == ROLE_NONE else value


def mask_from_indices(indices):
    mask = 0
    for index in indices:
        mask |= (1 << index)
    return mask


def indices_from_mask(mask):
    # Ascending index order -- reproduces the old "remaining players sorted"
    # canonicalisation for free (a bitmask is inherently order-free).
    return [i for i in range(8) if mask & (1 << i)]


def encode_team_permutation(remaining_mask, defender, attacker_a, attacker_b, discarded):
    return (remaining_mask
            | (_pack_role(defender) << _DEFENDER_SHIFT)
            | (_pack_role(attacker_a) << _ATTACKER_A_SHIFT)
            | (_pack_role(attacker_b) << _ATTACKER_B_SHIFT)
            | (_pack_role(discarded) << _DISCARDED_SHIFT))


def decode_team_permutation(code):
    """Return (remaining_mask, defender, attacker_A, attacker_B, discarded);
    roles are int index or None."""
    return (
        code & _REMAINING_MASK,
        _unpack_role(code, _DEFENDER_SHIFT),
        _unpack_role(code, _ATTACKER_A_SHIFT),
        _unpack_role(code, _ATTACKER_B_SHIFT),
        _unpack_role(code, _DISCARDED_SHIFT),
    )


def team_permutation_n(code):
    """Number of players on this team: popcount(remaining) + roles set. Mirrors
    the old TeamPermutation.get_n (remaining + defender + both attackers)."""
    remaining_mask, defender, attacker_a, attacker_b, _discarded = decode_team_permutation(code)
    n = bin(remaining_mask).count("1")
    if defender is not None:
        n += 1
    if attacker_a is not None:
        n += 1
    if attacker_b is not None:
        n += 1
    return n


def draft_stage_of_code(code):
    """Derive the draft stage from which roles are set, mirroring the old
    TeamPermutation.get_draft_stage."""
    _remaining_mask, defender, attacker_a, _attacker_b, discarded = decode_team_permutation(code)
    if discarded is not None:
        return DraftStage.discard_attacker
    elif attacker_a is not None:
        return DraftStage.select_attackers
    elif defender is not None:
        return DraftStage.select_defender
    else:
        return DraftStage.none
